src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -0,0 +1,477 @@
+        )
+
+        def promise_resolver(res: JsProxy, rej: JsProxy) -> None:
+            def onMsg(e: JsProxy) -> None:
+                self.streaming_ready = True
+                res(e)
+
+            def onErr(e: JsProxy) -> None:
+                rej(e)
+
+            self._worker.onmessage = onMsg
+            self._worker.onerror = onErr
+
+        dataURL = js.URL.createObjectURL(dataBlob)
+        self._worker = js.globalThis.Worker.new(dataURL)
+        self._worker_ready_promise = js.globalThis.Promise.new(promise_resolver)
+
+    def send(self, request: EmscriptenRequest) -> EmscriptenResponse:
+        headers = {
+            k: v for k, v in request.headers.items() if k not in HEADERS_TO_IGNORE
+        }
+
+        body = request.body
+        fetch_data = {"headers": headers, "body": to_js(body), "method": request.method}
+        # start the request off in the worker
+        timeout = int(1000 * request.timeout) if request.timeout > 0 else None
+        shared_buffer = js.SharedArrayBuffer.new(1048576)
+        int_buffer = js.Int32Array.new(shared_buffer)
+        byte_buffer = js.Uint8Array.new(shared_buffer, 8)
+
+        js.Atomics.store(int_buffer, 0, 0)
+        js.Atomics.notify(int_buffer, 0)
+        absolute_url = js.URL.new(request.url, js.location).href
+        self._worker.postMessage(
+            _obj_from_dict(
+                {
+                    "buffer": shared_buffer,
+                    "url": absolute_url,
+                    "fetchParams": fetch_data,
+                }
+            )
+        )
+        # wait for the worker to send something
+        js.Atomics.wait(int_buffer, 0, 0, timeout)
+        if int_buffer[0] == 0:
+            raise _TimeoutError(
+                "Timeout connecting to streaming request",
+                request=request,
+                response=None,
+            )
+        if int_buffer[0] == SUCCESS_HEADER:
+            # got response
+            # header length is in second int of intBuffer
+            string_len = int_buffer[1]
+            # decode the rest to a JSON string
+            decoder = js.TextDecoder.new()
+            # this does a copy (the slice) because decode can't work on shared array
+            # for some silly reason
+            json_str = decoder.decode(byte_buffer.slice(0, string_len))
+            # get it as an object
+            response_obj = json.loads(json_str)
+            return EmscriptenResponse(
+                request=request,
+                status_code=response_obj["status"],
+                headers=response_obj["headers"],
+                body=io.BufferedReader(
+                    _ReadStream(
+                        int_buffer,
+                        byte_buffer,
+                        request.timeout,
+                        self._worker,
+                        response_obj["connectionID"],
+                    ),
+                    buffer_size=1048576,
```

Comment:
Where does this value come from?

Reply:
It's an arbitrary value for thread buffer transfer that is a decently big buffer that many files will fit in, but not so big that it takes up tons of memory. Could be set as a constant somewhere though?